### PR TITLE
Setup a small toy example with an oauth2 flow

### PR DESCRIPTION
**Pull request by Jasopaum**

### CODE REVIEW

File: provider/src/main.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package main
```
Comment (simonvadee) [inline provider/src/main.go:1]: you should have a single file using "package main" (generally main.go).
Logic should be implemented in its own package

File: provider/src/main.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+// The session is a unique session identifier
+const sessionName = "authentication"
+
+// A state for performing the OAuth 2.0 flow. This is usually not part of a consent app, but in order for the demo
+// to make sense, it performs the OAuth 2.0 authorize code flow.
+var state = "demostatedemostatedemo"
+
+// Context we want handlers to have access to
+type Env struct {
```
Comment (simonvadee) [inline provider/src/main.go:22]: Env (and all its method eg: `func (e *Env) Myfunc() ...`) should be defined in their own file

File: provider/src/database.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
+package main
```
Comment (simonvadee) [inline provider/src/database.go:1]: you need a package specifically dedicated to user management

File: provider/src/database.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	Email    string `gorm:"type:varchar(100);unique_index"`
+	Name     string `gorm:"type:varchar(100)"`
+	Password string `json:"Password"`
+}
+
+type DB struct {
+	*gorm.DB
+}
+
+type UserStore interface {
```
Comment (simonvadee) [inline provider/src/database.go:29]: define the interface before defining the struct.
add comments

File: provider/src/database.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	AddUser(user *User) (*User, error)
+	FindUser(email string, password string) (*User, error)
+}
+
+func ConnectDB() *DB {
```
Comment (simonvadee) [inline provider/src/database.go:34]: by convention, function that create structs are called `NewStructName()` eg `NewDB()`

File: provider/src/database.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		databaseHost, port, username, databaseName, password,
+	)
+
+	db, err := gorm.Open("postgres", dbURI)
+
+	if err != nil {
+		fmt.Println("error", err)
+		panic(err)
+	}
+	// defer db.Close()
```
Comment (simonvadee) [inline provider/src/database.go:52]: yeah this shouldn't be here

File: provider/src/database.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	// TODO check doc about that
+	db.AutoMigrate(&User{})
+
+	return &DB{db}
+}
+
+func (db *DB) FindUser(email string, password string) (*User, error) {
+	user := &User{}
+
+	if err := db.Where("Email = ?", email).First(user).Error; err != nil {
```
Comment (simonvadee) [inline provider/src/database.go:64]: do you have to use a string to make the query here ? can't you use your model's attribute user.Email ?

Comment (simonvadee) [reply provider/src/database.go:64]: I prefer this:
`db.Where(&User{Name: "jinzhu", Age: 20}).First(&user)`
what do you think ?

File: provider/src/consent.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+	params.Add("consent_challenge", challenge)
+
+	getUrl := fmt.Sprintf("%s?%s", env.hConf.ConsentRequestRoute, params.Encode())
+	resp, _ := http.Get(getUrl)
+
+	b, _ := ioutil.ReadAll(resp.Body)
+	var jsonResp struct {
+		RequestedScopes []string `json:"requested_scope"`
+	}
+	json.Unmarshal(b, &jsonResp)
```
Comment (simonvadee) [inline provider/src/consent.go:30]: iferr

File: provider/src/consent.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+func (env *Env) handleConsent(w http.ResponseWriter, r *http.Request) {
+	// Get the consent requerst id from the query.
+	challenge := r.URL.Query().Get("consent_challenge")
+
+	// Get info about current flow
+	params := url.Values{}
+	params.Add("consent_challenge", challenge)
+
+	getUrl := fmt.Sprintf("%s?%s", env.hConf.ConsentRequestRoute, params.Encode())
+	resp, _ := http.Get(getUrl)
```
Comment (simonvadee) [inline provider/src/consent.go:24]: iferr

File: provider/src/main.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+	hConf := &hydraConfig{
+		LoginRequestRoute:   "http://localhost:4445/oauth2/auth/requests/login",
+		ConsentRequestRoute: "http://localhost:4445/oauth2/auth/requests/consent",
+	}
+	db := ConnectDB()
+
+	envContext := &Env{hConf, db}
+
+	// Set up a router and some routes
+	http.HandleFunc("/login", envContext.handleLogin)
```
Comment (simonvadee) [inline provider/src/main.go:38]: I would advise you not to use the native http package. I personally like [this router](github.com/julienschmidt/httprouter) (which is compatible with native http package). For instance, you can define GET or POST routes directly (istead of an ugly if request.Method == "POST" in hour handler)
You can find a nice example in river-api :)

File: provider/src/users.go [hunk 11]
```diff
@@ -0,0 +1,19 @@
+package main
+
+import (
+	"encoding/json"
+	"net/http"
+)
+
+func (env *Env) handleSignup(w http.ResponseWriter, r *http.Request) {
+	user := &User{}
+	json.NewDecoder(r.Body).Decode(user)
```
Comment (simonvadee) [inline provider/src/users.go:10]: iferr

File: provider/src/main.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+import (
+	"log"
+	"net/http"
+
+	"github.com/gorilla/sessions"
+	"github.com/ory/common/env"
+)
+
+// This store will be used to save user authentication
+var store = sessions.NewCookieStore([]byte("something-very-secret-keep-it-safe"))
```
Comment (simonvadee) [inline provider/src/main.go:12]: do we want to persist this ?

File: provider/src/main.go (hunk 2 above)
Comment (simonvadee) [reply provider/src/main.go:22]: also, `Env` is to vague (and has many meanings), maybe AuthContext or something ?

File: provider/src/users.go [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+	createdUser, err := env.db.AddUser(user)
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+
+	json.NewEncoder(w).Encode(createdUser)
```
Comment (simonvadee) [inline provider/src/users.go:18]: iferr

File: provider/templates/consent.html [hunk 14]
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE html>
```
Comment (simonvadee) [inline provider/templates/consent.html:1]: i'm sure this is some sexy UI

Comment (Jasopaum) [reply provider/templates/consent.html:1]: It's perfect ❤️

File: provider/src/main.go (hunk 12 above)
Comment (Jasopaum) [reply provider/src/main.go:12]: Not sure, I don't know why we'd need that in the provider but I'm still not an authentication expert so... 🤷‍♂️ 
I think we could remove it

File: provider/src/database.go (hunk 6 above)
Comment (Jasopaum) [reply provider/src/database.go:52]: Yep!
We'd never want to close the DB connection in this service, right?

File: provider/src/database.go (hunk 7 above)
Comment (Jasopaum) [reply provider/src/database.go:64]: Agreed, the syntax with strings is ugly, I haven't read gorm's doc yet, I wanted to do that next 🙂

File: provider/src/main.go (hunk 1 above)
Comment (Jasopaum) [reply provider/src/main.go:1]: What about the http handlers?
So we could have the packages "users", "context", "main", maybe another one?

Also, what do you think about the types.go file? Not sure it's very idiomatic

File: provider/src/database.go (hunk 4 above)
Comment (Jasopaum) [reply provider/src/database.go:29]: Is that a general convention rule? When defining interfaces and structs, define interfaces first?

File: provider/src/main.go (hunk 2 above)
Comment (Jasopaum) [reply provider/src/main.go:22]: I agree on the naming
For the files, I didn't really want to put all the router functions to go to the same place. Maybe they can go to different files but in the same package? But it seems odd to me to have them in a side package when they really are the core of what's being done here

File: provider/src/database.go (hunk 4 above)
Comment (simonvadee) [reply provider/src/database.go]: well thats what I'm used to, yes. There might be many implementations of an interface so start by defining the interface seems logical to me

File: provider/cmd/provider/main.go [hunk 15]
```diff
@@ -5,7 +5,10 @@ import (
 	"net/http"
 
 	"github.com/gorilla/sessions"
+	"github.com/julienschmidt/httprouter"
 	"github.com/ory/common/env"
+
+	"provider/internal/users"
```
Comment (simonvadee) [inline provider/cmd/provider/main.go:11]: generally in go packages I avoid "src" "internal" "lib" etc... folders. I think I would go for
identity-provider/main.go

identity-provider/provider/provider.go
identity-provider/provider/login.go
identity-provider/provider/signup.go
identity-provider/provider/consent.go
identity-provider/provider/types.go
identity-provider/provider/ui/*.html

identity-provider/users/user.go
 
identity-provider/example-client
identity-provider/example-client/main.go
identity-provider/example-client/templates/*.html

File: provider/cmd/provider/login.go [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+	// // Let's create a session where we store the user id. We can ignore errors from the session store
+	// // as it will always return a session!
+	// session, _ := store.Get(r, sessionName)
+	// session.Values["user"] = "buzz-lightyear"
+
+	// // Store the session in the cookie
+	// if err := store.Save(r, w, session); err != nil {
+	// 	http.Error(w, errors.Wrap(err, "Could not persist cookie").Error(), http.StatusBadRequest)
+	// 	return
+	// }
```
Comment (simonvadee) [inline provider/cmd/provider/login.go:70]: we don't need this ?

File: provider/cmd/provider/consent.go [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+
+	"github.com/julienschmidt/httprouter"
+	"github.com/pkg/errors"
+)
+
+// After pressing "click here", the Authorize Code flow is performed and the user is redirected to Hydra. Next, Hydra
+// validates the consent request (it's not valid yet) and redirects us to the consent endpoint which we set with `CONSENT_URL=http://localhost:4445/consent`.
+func (env *Env) getConsent(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	// Get the consent requerst id from the query.
+	challenge := r.URL.Query().Get("consent_challenge")
```
Comment (simonvadee) [inline provider/cmd/provider/consent.go:17]: what happens if this is missing from the url query ?

File: provider/cmd/provider/consent.go [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+	requestedScopes := jsonResp.RequestedScopes
+	// TODO do stuff with response
+
+	// This helper checks if the user is already authenticated. If not, we
+	// redirect them to the login endpoint.
+	// user := authenticated(r)
+	// if user == "" {
+	// 	http.Redirect(w, r, "/login?consent="+consentRequestID, http.StatusFound)
+	// 	return
+	// }
```
Comment (simonvadee) [inline provider/cmd/provider/consent.go:48]: don't need this ?

File: provider/src/main.go (hunk 1 above)
Comment (simonvadee) [reply provider/src/main.go]: see my comment below about folder structure

Comment (simonvadee) [reply provider/src/main.go]: about types.go it depends: if your types are used across multiple files (which I find to be rarely the case in go), it's of to have a file dedicated to your types. However, if you add method on your structs, they should have their own file.

File: provider/src/main.go (hunk 2 above)
Comment (simonvadee) [reply provider/src/main.go]: same package different files sounds good. If you have a package "provider" at the root of your repo, it's not a "side" package imho.

File: provider/src/database.go (hunk 6 above)
Comment (simonvadee) [reply provider/src/database.go]: well we could close it if the http server fails to launch, or on ctrl+c for instance

File: provider/src/main.go (hunk 12 above)
Comment (simonvadee) [reply provider/src/main.go]: can we still have cookies if we remove that ?

Comment (Jasopaum) [reply provider/src/main.go]: I don't know. Do you see why we'd need cookies in the provider?

Comment (simonvadee) [reply provider/src/main.go]: no I don't, but my understanding is not good enough..

**Pull request merged**